README.md

```diff
@@ -206,36 +124,63 @@ assert.isFalse(result.success, result.error);
 ```
 
 ## Contracts with State
-sCrypt offers [stateful contracts](https://medium.com/xiaohuiliu/stateful-smart-contracts-on-bitcoin-sv-c24f83a0f783). `OP_RETURN` data of the contract locking script can be accessed by using an accessor named `dataPart`, for example:
+sCrypt offers [stateful contracts](https://scryptdoc.readthedocs.io/en/latest/state.html#stateful-contract). Declare any property that is part of the state with a decorator `@state` in sCrypt contract, for example:
 ```typescript
-const dataPart = instance.dataPart;
-const dataPartASM = instance.dataPart.toASM();
-const dataPartHex = instance.dataPart.toHex();
-// to set it using ASM
-instance.setDataPart(dataInASM);
-// to set it using state object (no nesting)
-let state = {'counter': 11, 'bytes': '1234', 'flag': true}
-instance.setDataPart(state)
+contract Counter {
+    @state
+    int counter;
+
+    constructor(int counter) {
+        this.counter = counter;
+    }
+}
 ```
-After that, the `instance.lockingScript` would include the data part automatically.
 
-If you want to access the code part of the contract's locking script including the trailing `OP_RETURN`, use:
+Use the initial state to instantiate the contract.
+
 ```typescript
-const codePart = instance.codePart;
-const codePartASM = instance.codePart.toASM();
-const codePartHex = instance.codePart.toHex();
+const instance = new Counter(1000);
 ```
 
-Another way to access the state is to use the `state` decorator in the sCrypt contract. 
 Then you can read and write the state by accessing the properties of the contract instance.
 
+
 ```typescript
+// read state
+let state = instance.counter;
+
 // update state
-stateExample.counter++;
-stateExample.state_bytes = new Bytes('010101');
-stateExample.state_bool = false;
+instance.counter++;
 ```
 
+Finally, you can use `lockingScript` to build transaction output and use `prevLockingScript` to get Sighash Preimage of the transaction.
+
+```typescript
+tx.addOutput(new bsv.Transaction.Output({
+  script: instance.lockingScript,
+  satoshis: outputAmount
+}))
+
+preimage = getPreimage(tx, instance.prevLockingScript, inputSatoshis)
+
+// set txContext for verification
+instance.txContext = {
+  tx,
+  inputIndex,
+  inputSatoshis
+}
+```
+
+If you want to update the states for the next time, you need to save the latest states of the contract by calling `commitState`.
+
+```typescript
+//Save updated states
+instance.commitState();
+```
+
+Another way to maintain state is to directly access the OP_RETURN data of the contract through [dataPart](docs/DATAPART.md)
```

Comment:
You can also maintain state manually to, for example, optimize your contract or use customized state de/serialization [rawstate](docs/RAWSTATE.md)